### DISCUSSION TITLE
chapter_stack_and_queue/summary/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#58] [Announcements]
# chapter_stack_and_queue/summary/

Your first book to learn Data Structure And Algorithm.

https://www.hello-algo.com/chapter_stack_and_queue/summary/

**Comment by feiaiyue** [comment]
K神！！～能否写个优先队列呢
还有会写一些图论的算法吗

**Reply by krahets** [reply]
必须的，加足马力写！～

**Comment by muzishuiji** [comment]
浏览器的「前进」与「后退」是不是和撤销（undo）和反撤销（redo）同理，使用两个堆栈实现的？

**Reply by krahets** [reply]
是的，前进+后退和撤销+反撤销是一回事

**Comment by showlibia** [comment]
DAY4，双向队列里实现「出入队列」的方法加入 **isFront** 的判断，这样的tips很妙啊

**Reply by krahets** [reply]
哈哈，这样写可以缩短一些代码，并且清晰地展示头插和尾插在代码上的不同处。不过在实际中，分开到两个函数写可能更常见，代码更清晰，可读性更好。

**Comment by Chris-Ran** [comment]
请问“如果后续仍需要使用弹出节点，则不需要释放内存。若之后不需要用到，Java 和 Python 等语言拥有自动垃圾回收机制，因此不需要手动释放内存；在 C 和 C++ 中需要手动释放内存。“中，C++手动释放栈道内存在代码中怎么体现呢？直接delete stack吗？

**Reply by krahets** [reply]
Hi，需要释放每个节点的内存，这部分可以写在析构函数中：

```cpp
/* 释放链表节点内存 */
void freeMemoryLinkedList(ListNode *cur) {
    // 释放内存
    ListNode *pre;
    while (cur != nullptr) {
        pre = cur;
        cur = cur->next;
        delete pre;
    }
}
```

然后再 `delete stack;` 。

**Comment by HuaWang135608** [comment]
Over

**Comment by Chr1sdian** [comment]
Day5

**Comment by Dannytaotao** [comment]
5.14 finish

**Comment by DayDayYES** [comment]
栈和队列都可以由数组或者链表形成，数组形成栈的栈顶在上方，栈底在下方，而数组头在下面，数组尾在上面。链表的表头对应栈头，链尾对应栈尾。数组形成的队列的头指针指向队首元素，尾指针指向队尾元素的下一个位置，链表形成的队列头指针指向头节点，尾指针指向尾节点。队列入队是从队尾进，栈入栈是从栈顶进

**Comment by fakerUNcode** [comment]
现代浏览器应该主要使用双栈结构，具体的操作流程如下：

访问新页面：
当前页面压入“后退栈”。
清空“前进栈”以防止出现不一致的记录（因为访问新页面意味着“前进”的历史无效）。
新页面成为当前页面。

点击后退：
当前页面压入“前进栈”。
从“后退栈”弹出栈顶页面，并将其设为当前页面。

点击前进：
当前页面压入“后退栈”。
从“前进栈”弹出栈顶页面，并将其设为当前页面。

**Reply by fakerUNcode** [reply]
这个应该和undo\redo是类似的操作方式

**Comment by xiaozhuge123** [comment]
day04

**Comment by solink7** [comment]
# 打卡 2025/6/13

**Comment by yyym2017** [comment]
### 试试 markdown

**Comment by sunshinesDL** [comment]
对于问题 `Q：撤销（undo）和反撤销（redo）具体是如何实现的？`，

回答中的步骤一指出 `当用户执行一个操作，将这个操作压入栈 A ，并清空栈 B 。`，起初我不明白为啥将新操作压入栈 A 后，要立即清空栈 B ？

思考后，我认为这样做是为了维护前后操作的正确顺序。否则，如果将新操作压入栈 A 后不清空栈 B，那么此时若启用“反撤销”功能把旧操作加载回栈 A，就会导致旧操作比当前操作更“新”，从而打乱了前后操作的原有顺序。

**Comment by mako-kawai** [comment]
***2026.3.21***
